controllers/spacebindingcleanup/spacebinding_cleanup_controller_test.go

```diff
@@ -26,29 +26,19 @@ import (
 
 	"github.com/stretchr/testify/require"
 	"k8s.io/client-go/kubernetes/scheme"
-	"sigs.k8s.io/controller-runtime/pkg/client"
 	runtimeclient "sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 )
 
-type publicViewerConfigMock struct{ enabled bool }
-
-func (m publicViewerConfigMock) Enabled() bool {
-	return m.enabled
+func TestDeleteSpaceBindingWithPublicViewerDisabled(t *testing.T) {
+	testDeleteSpaceBinding(t, false)
 }
 
-func newPublicViewerConfigProvider(enabled bool) PublicViewerConfigProvider {
-	return func(context.Context, client.Client) PublicViewerConfig {
-		return publicViewerConfigMock{enabled: enabled}
-	}
+func TestDeleteSpaceBindingWithPublicViewerEnabled(t *testing.T) {
+	testDeleteSpaceBinding(t, true)
 }
 
-var (
-	publicViewerConfigEnabledMockProvider  PublicViewerConfigProvider = newPublicViewerConfigProvider(true)
-	publicViewerConfigDisabledMockProvider PublicViewerConfigProvider = newPublicViewerConfigProvider(false)
-)
-
-func TestDeleteSpaceBinding(t *testing.T) {
+func testDeleteSpaceBinding(t *testing.T, publicViewerEnabled bool) {
```

Comment:
is there any reason for executing the tests twice , once with the publicviewer config disabled and one enabled ? 

Or we could add a sub test section covering this new scenario in the existing tests, something like:

<details>

```

func TestDeleteSpaceBinding(t *testing.T) {
	// given
	sbLaraRedhatAdmin := sb.NewSpaceBinding("lara", "redhat", "admin", "signupA")
	sbJoeRedhatView := sb.NewSpaceBinding("joe", "redhat", "view", "signupB")
	sbLaraIbmEdit := sb.NewSpaceBinding("lara", "ibm", "edit", "signupC")

	redhatSpace := spacetest.NewSpace(test.HostOperatorNs, "redhat")
	ibmSpace := spacetest.NewSpace(test.HostOperatorNs, "ibm")

	laraMur := masteruserrecord.NewMasterUserRecord(t, "lara")
	joeMur := masteruserrecord.NewMasterUserRecord(t, "joe")

	t.Run("lara-redhat SpaceBinding removed when redhat space is missing", func(t *testing.T) {

		fakeClient := test.NewFakeClient(t, sbLaraRedhatAdmin, sbJoeRedhatView, sbLaraIbmEdit, laraMur, joeMur, ibmSpace)
		reconciler := prepareReconciler(t, fakeClient)

		// when
		res, err := reconciler.Reconcile(context.TODO(), requestFor(sbLaraRedhatAdmin))

		// then
		require.Equal(t, res.RequeueAfter, time.Duration(0)) // no requeue
		require.NoError(t, err)
		sb.AssertThatSpaceBinding(t, test.HostOperatorNs, "lara", "redhat", fakeClient).DoesNotExist()
		sb.AssertThatSpaceBinding(t, test.HostOperatorNs, "joe", "redhat", fakeClient).Exists()
		sb.AssertThatSpaceBinding(t, test.HostOperatorNs, "lara", "ibm", fakeClient).Exists()
	})

	t.Run("joe-redhat SpaceBinding removed when joe MUR is missing", func(t *testing.T) {

		fakeClient := test.NewFakeClient(t, sbJoeRedhatView, sbLaraRedhatAdmin, sbLaraIbmEdit, laraMur, ibmSpace, redhatSpace)
		reconciler := prepareReconciler(t, fakeClient)

		// when
		res, err := reconciler.Reconcile(context.TODO(), requestFor(sbJoeRedhatView))

		// then
		require.Equal(t, res.RequeueAfter, time.Duration(0)) // no requeue
		require.NoError(t, err)
		sb.AssertThatSpaceBinding(t, test.HostOperatorNs, "lara", "redhat", fakeClient).Exists()
		sb.AssertThatSpaceBinding(t, test.HostOperatorNs, "joe", "redhat", fakeClient).DoesNotExist()
		sb.AssertThatSpaceBinding(t, test.HostOperatorNs, "lara", "ibm", fakeClient).Exists()
	})

	t.Run("lara-redhat SpaceBinding is being deleted, so no action needed", func(t *testing.T) {
		sbLaraRedhatAdmin := sbLaraRedhatAdmin.DeepCopy()
		now := metav1.Now()
		sbLaraRedhatAdmin.DeletionTimestamp = &now
		fakeClient := test.NewFakeClient(t, sbLaraRedhatAdmin, sbJoeRedhatView, sbLaraIbmEdit, laraMur, joeMur, ibmSpace)
		reconciler := prepareReconciler(t, fakeClient)

		// when
		res, err := reconciler.Reconcile(context.TODO(), requestFor(sbLaraRedhatAdmin))

		// then
		require.Equal(t, res.RequeueAfter, time.Duration(0)) // no requeue
		require.NoError(t, err)
		sb.AssertThatSpaceBinding(t, test.HostOperatorNs, "lara", "redhat", fakeClient).Exists()
		sb.AssertThatSpaceBinding(t, test.HostOperatorNs, "joe", "redhat", fakeClient).Exists()
		sb.AssertThatSpaceBinding(t, test.HostOperatorNs, "lara", "ibm", fakeClient).Exists()
	})

	t.Run("no SpaceBinding removed when MUR and Space are present", func(t *testing.T) {

		fakeClient := test.NewFakeClient(t, sbLaraRedhatAdmin, sbJoeRedhatView, sbLaraIbmEdit, laraMur, joeMur, ibmSpace, redhatSpace)
		reconciler := prepareReconciler(t, fakeClient)

		// when
		res, err := reconciler.Reconcile(context.TODO(), requestFor(sbLaraRedhatAdmin))

		// then
		require.Equal(t, res.RequeueAfter, time.Duration(0)) // no requeue
		require.NoError(t, err)
		sb.AssertThatSpaceBinding(t, test.HostOperatorNs, "lara", "redhat", fakeClient).Exists()
		sb.AssertThatSpaceBinding(t, test.HostOperatorNs, "joe", "redhat", fakeClient).Exists()
		sb.AssertThatSpaceBinding(t, test.HostOperatorNs, "lara", "ibm", fakeClient).Exists()
	})

	t.Run("fails while getting the bound resource", func(t *testing.T) {

		for _, boundResourceName := range []string{"lara", "redhat"} {
			fakeClient := test.NewFakeClient(t, sbLaraRedhatAdmin, redhatSpace, laraMur)
			reconciler := prepareReconciler(t, fakeClient)
			fakeClient.MockGet = func(ctx context.Context, key runtimeclient.ObjectKey, obj runtimeclient.Object, opts ...runtimeclient.GetOption) error {
				if key.Name == boundResourceName {
					return fmt.Errorf("some error")
				}
				return fakeClient.Client.Get(ctx, key, obj, opts...)
			}

			// when
			_, err := reconciler.Reconcile(context.TODO(), requestFor(sbLaraRedhatAdmin))

			// then
			require.Error(t, err)
			sb.AssertThatSpaceBinding(t, test.HostOperatorNs, "lara", "redhat", fakeClient).Exists()
		}
	})

	t.Run("fails while deleting the SpaceBinding", func(t *testing.T) {

		fakeClient := test.NewFakeClient(t, sbLaraRedhatAdmin, redhatSpace)
		reconciler := prepareReconciler(t, fakeClient)
		fakeClient.MockDelete = func(ctx context.Context, obj runtimeclient.Object, opts ...runtimeclient.DeleteOption) error {
			return fmt.Errorf("some error")
		}

		// when
		_, err := reconciler.Reconcile(context.TODO(), requestFor(sbLaraRedhatAdmin))

		// then
		require.Error(t, err)
		sb.AssertThatSpaceBinding(t, test.HostOperatorNs, "lara", "redhat", fakeClient).Exists()
	})

	t.Run("with public viewer enabled ", func(t *testing.T) {
		sbPublicViewerRedhatView := sb.NewSpaceBinding(toolchainv1alpha1.KubesawAuthenticatedUsername, "redhat", "view", "signupD")
		toolchainconfig := commonconfig.NewToolchainConfigObjWithReset(t, testconfig.PublicViewerConfig(true))

		t.Run("kubesaw-authenticated SpaceBinding is removed when redhat space is missing", func(t *testing.T) {
			fakeClient := test.NewFakeClient(t, sbPublicViewerRedhatView, toolchainconfig)
			reconciler := prepareReconciler(t, fakeClient)

			// when
			res, err := reconciler.Reconcile(context.TODO(), requestFor(sbPublicViewerRedhatView))

			// then
			require.Equal(t, res.RequeueAfter, time.Duration(0)) // no requeue
			require.NoError(t, err)
			sb.AssertThatSpaceBinding(t, test.HostOperatorNs, toolchainv1alpha1.KubesawAuthenticatedUsername, "redhat", fakeClient).DoesNotExist()
		})

		t.Run("kubesaw-authenticated SpaceBinding is NOT removed when MUR is missing", func(t *testing.T) {
			fakeClient := test.NewFakeClient(t, redhatSpace, sbPublicViewerRedhatView, toolchainconfig)
			reconciler := prepareReconciler(t, fakeClient)

			// when
			res, err := reconciler.Reconcile(context.TODO(), requestFor(sbPublicViewerRedhatView))

			// then
			require.Equal(t, res.RequeueAfter, time.Duration(0)) // no requeue
			require.NoError(t, err)
			sb.AssertThatSpaceBinding(t, test.HostOperatorNs, toolchainv1alpha1.KubesawAuthenticatedUsername, "redhat", fakeClient).Exists()
		})
	})
}
```
</details>

Reply:
we want to test that the behavior the controller has when public-viewer is disabled is the same it has when public-viewer is enabled, withthe exception of two following tests:
* kubesaw-authenticated SpaceBinding is removed when redhat space is missing
* kubesaw-authenticated SpaceBinding is NOT removed when MUR is missing

In other words, we are testing that the behavior of the controller remains the same on SpaceBindings that are not related to kubesaw-authenticated when public-viewer is enabled. I'd say it would be good to move the two runs wrapped in the if at [L55](https://github.com/codeready-toolchain/host-operator/blob/ce0e5e67deccdf92bea8bf738b43348aaf76ac6f/controllers/spacebindingcleanup/spacebinding_cleanup_controller_test.go#L55) to a dedicated Test function

Reply:
Thanks for looking into this and for moving the public-viewer use case into a dedicated tests!

Reply:
But still the way I see it, it's that you need to add 2 specific uses cases within the existing tests:
- `TestDeleteSpaceBinding` which checks that dangling SpaceBindings are deleted as expected
- `TestDeleteSpaceBindingRequest` which checks that SpaceBindingRequest generated from SBRs are deleted as expected

by changing the name of the tests, you are giving those a new perspective , which is now focused on the public viewer feature. 

Let's say tomorrow we have a new feature `BlaBla` that needs some special handling as well in this cleanup controller, are we going to : 
- modify the test names again and make it specific like: `TestCommonDeleteSpaceBindingWithPublicViewerDisabledAndBlaBlaEnabled` and same for SpaceRequest test
OR 
- just add a subtests in both SpaceBinding and SpaceBindingRequest tests that has a setup configured as we expect this new feature to be configured in prod?  I would do the later but this is just how I see it.

Reply:
Are you saying that by doing this we are also checking that the tests are still passing when the public viewer feature is enabled ?

Reply:
Great analysis, thanks for looking into it. I don't like the naming too and I agree that this can be a problem in case we introduce a new feature. Having a better naming would be more than appropriate.

What I think we should test here -and I tried to implement- is the following:
* **normal behavior (i.e., public-viewer disabled)**: the host-operator deletes a SpaceBinding (in the following SB) if the owner's MasterUserRecord (MUR) is not found
* **public-viewer enabled**: the host-operator deletes a SB if the owner's MUR is not found <ins>unless the MUR is `kubesaw-authenticated`</ins>

Hence, we can reuse the same set of tests. The only different behavior would be when the MUR is `kubesaw-authenticated`. In the former scenario the SBs for `kubesaw-authenticated` are deleted (that's missing BTW, I'll add these tests), whereas in the latter they are not deleted.

Any idea on how to organize the tests? 🙏🏾 😄

Reply:
WDYT something like this:

```go
func TestDeleteSpaceBinding(t *testing.T) {
	testCases := []struct {
		toolchainconfig *toolchainv1alpha1.ToolchainConfig
	}{
		{toolchainconfig: commonconfig.NewToolchainConfigObjWithReset(t, testconfig.PublicViewerConfig(false))},
		{toolchainconfig: commonconfig.NewToolchainConfigObjWithReset(t, testconfig.PublicViewerConfig(true))},
	}

	for _, tc := range testCases {
		toolchainconfig := tc.toolchainconfig

		// given
		sbLaraRedhatAdmin := sb.NewSpaceBinding("lara", "redhat", "admin", "signupA")
		sbJoeRedhatView := sb.NewSpaceBinding("joe", "redhat", "view", "signupB")
		sbLaraIbmEdit := sb.NewSpaceBinding("lara", "ibm", "edit", "signupC")

```

Reply:
I implemented that in 565d59e888bcdac851f5286a0686a8c780e457ef